src/sysupdate/sysupdated.c

```diff
@@ -0,0 +1,1895 @@
+                if (r < 0)
+                        return r;
+        }
+
+        r = sd_bus_message_new_method_return(msg, &reply);
+        if (r < 0)
+                return r;
+
+        r = sd_bus_message_append_strv(reply, urls);
+        if (r < 0)
+                return r;
+
+        return sd_bus_send(NULL, reply, NULL);
+}
+
+static const sd_bus_vtable manager_vtable[] = {
+        SD_BUS_VTABLE_START(0),
+
+        SD_BUS_METHOD_WITH_ARGS("ListTargets",
+                                SD_BUS_NO_ARGS,
+                                SD_BUS_RESULT("a(sso)", targets),
+                                method_list_targets,
+                                SD_BUS_VTABLE_UNPRIVILEGED),
+
+        SD_BUS_METHOD_WITH_ARGS("ListJobs",
+                                SD_BUS_NO_ARGS,
+                                SD_BUS_RESULT("a(usuo)", jobs),
+                                method_list_jobs,
+                                SD_BUS_VTABLE_UNPRIVILEGED),
+
+        SD_BUS_METHOD_WITH_ARGS("ListAppStream",
+                                SD_BUS_NO_ARGS,
+                                SD_BUS_RESULT("as", urls),
+                                method_list_appstream,
+                                SD_BUS_VTABLE_UNPRIVILEGED),
+
+        SD_BUS_SIGNAL_WITH_ARGS("JobRemoved",
+                                SD_BUS_ARGS("u", id, "o", path, "i", status),
+                                0),
+
+        SD_BUS_VTABLE_END
+};
+
+static const BusObjectImplementation manager_object = {
+        "/org/freedesktop/sysupdate1",
+        "org.freedesktop.sysupdate1.Manager",
+        .vtables = BUS_VTABLES(manager_vtable),
+        .children = BUS_IMPLEMENTATIONS(&job_object, &target_object),
+};
+
+static int manager_add_bus_objects(Manager *m) {
+        int r;
+
+        assert(m);
+
+        r = bus_add_implementation(m->bus, &manager_object, m);
+        if (r < 0)
+                return r;
+
+        r = bus_log_control_api_register(m->bus);
+        if (r < 0)
+                return r;
+
+        r = sd_bus_request_name_async(m->bus, NULL, "org.freedesktop.sysupdate1", 0, NULL, NULL);
+        if (r < 0)
+                return log_error_errno(r, "Failed to request name: %m");
+
+        r = sd_bus_attach_event(m->bus, m->event, 0);
+        if (r < 0)
+                return log_error_errno(r, "Failed to attach bus to event loop: %m");
+
+        return 0;
+}
+
+static bool manager_idle(void *userdata) {
+        Manager *m = ASSERT_PTR(userdata);
+
+        if (!hashmap_isempty(m->jobs))
+                return false;
+
+        /* The daemon is completely idle, and we will exit in ~30s as long as
+         * no new clients appear. This is the only time we can safely clear
+         * out our targets, so that when a new client appears it will see
+         * up-to-date information.
+         */
+        hashmap_clear(m->targets);
+        log_debug("Cleared target cache");
```

Comment:
please move that into a separate idle handler. the bus idle stuff should not be reused for that, because we nowadays make exit-on-idle optional

Reply:
or in other words, this really should just check if there are ongoing jobs and nothing else

Reply:
I moved this into job_free, removing the target cache when the last job gets removed, although I'm not sure this was the best approach.

Reply:
That works. Just from a code-readability standpoint, however, I don't like that it's unclear what exactly is going on there. I'd define either a `manager_maybe_became_idle()` (that checks m->jobs, then clears the target cache) or a `manager_became_idle()` (that just clears the target cache, and you'd check m->jobs at the call site). Either function would be defined close to the other idle-handling code. Also, this gives an opportunity to preserve the comment & logging